crates/red_knot_workspace/tests/check.rs

```diff
@@ -28,17 +33,116 @@ fn setup_db(workspace_root: SystemPathBuf) -> anyhow::Result<RootDatabase> {
+struct PullTypesVisitor<'db> {
+    model: SemanticModel<'db>,
+}
+
+impl<'db> PullTypesVisitor<'db> {
+    fn new(db: &'db RootDatabase, file: File) -> Self {
+        Self {
+            model: SemanticModel::new(db, file),
+        }
+    }
+}
+
+impl SourceOrderVisitor<'_> for PullTypesVisitor<'_> {
+    fn visit_expr(&mut self, expr: &Expr) {
+        let _ty = expr.ty(&self.model);
+
+        source_order::walk_expr(self, expr);
+    }
+
+    fn visit_alias(&mut self, alias: &Alias) {
+        let _ty = alias.ty(&self.model);
+
+        source_order::walk_alias(self, alias);
+    }
+
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::FunctionDef(function) => {
+                let _ty = function.ty(&self.model);
+            }
+            Stmt::ClassDef(class) => {
+                let _ty = class.ty(&self.model);
+            }
+            Stmt::AnnAssign(assign) => {
+                let _assignment_ty = assign.ty(&self.model);
+
+                // FIXME: I think the following constraint should be true in all cases but it isn't.
+                // let target_ty = assign.target.ty(&self.model);
+                //
+                // assert_eq!(_assignment_ty, target_ty);
+            }
+            Stmt::Return(_)
+            | Stmt::Delete(_)
+            | Stmt::Assign(_)
+            | Stmt::AugAssign(_)
+            | Stmt::TypeAlias(_)
+            | Stmt::For(_)
+            | Stmt::While(_)
+            | Stmt::If(_)
+            | Stmt::With(_)
+            | Stmt::Match(_)
+            | Stmt::Raise(_)
+            | Stmt::Try(_)
+            | Stmt::Assert(_)
+            | Stmt::Import(_)
+            | Stmt::ImportFrom(_)
+            | Stmt::Global(_)
+            | Stmt::Nonlocal(_)
+            | Stmt::Expr(_)
+            | Stmt::Pass(_)
+            | Stmt::Break(_)
+            | Stmt::Continue(_)
+            | Stmt::IpyEscapeCommand(_) => {}
+        }
+
+        source_order::walk_stmt(self, stmt);
+    }
+
+    fn visit_comprehension(&mut self, comprehension: &Comprehension) {
+        let _ty = comprehension.ty(&self.model);
+        source_order::walk_comprehension(self, comprehension);
+    }
+
+    fn visit_parameter(&mut self, parameter: &Parameter) {
+        let _ty = parameter.ty(&self.model);
+
+        source_order::walk_parameter(self, parameter);
+    }
+
+    fn visit_parameter_with_default(&mut self, parameter_with_default: &ParameterWithDefault) {
+        let _ty = parameter_with_default.ty(&self.model);
+
+        // FIXME: We currently don't create a definition for the nested parameter in the semantic builder.
+        //   This seems correct to me because we otherwise end up with two definitions for the same symbol.
+        //   However, it breaks the contract that `parameter.ty` always returns a type (doesn't panic).
+        //   Not sure what the right fix is, but it's out of scope for changing the tests.
+        //   Example: `def foo(bar): ...` panics
```

Comment:
I think the difference here is that for parameters this is actually useful. For example: give me the symbol of the parameter and what is its declared type is an operation that seems useful to me. I definitely don't want partial HasTy implementations that sometimes panic

Reply:
I think one way to make this not panic, without changes to the AST, might be to do scope-wide type inference (and make sure that sets types on all parameter nodes), and fallback to that route for parameter nodes that don't have an entry in the node-to-definitions map in semantic index.

Or it might also be possible to handle this by putting two entries into the node-to-definitions map (one for the `ParameterWithDefault` and another for the nested `Parameter`) that both map to the same Definition. This would effectively give us a very localized way to "walk up" that one level of the AST. Not sure if there'd be any downsides to having two nodes map to the same Definition in that map.

Reply:
That's a great suggestion. I implemented the definition redirect.